File: 백준/Gold/10026. 적록색약/적록색약.py
'''
아이디어:
bfs를 통한 풀이
boolean grpah로 만들어서 쉽게 풀기
'''

import sys
from collections import deque

def is_valid(x,y):
    return 0<=x<N and 0<=y<N

delta_x = [0,0,-1,1]
delta_y = [1,-1,0,0]

normal_graph = []
N = int(sys.stdin.readline())
for _ in range(N):
    normal_graph.append(list(sys.stdin.readline().rstrip()))

blind_graph = []
for i in range(N):
    blind_graph.append([])
    for j in range(N):
        if normal_graph[i][j] == 'B':
            blind_graph[i].append(False)
        else:
            blind_graph[i].append(True)

visited = []
for _ in range(N):
    visited.append([False]*N)

cnt_normal = 0
for i in range(N):
    for j in range(N):
        if not visited[i][j]:
            cnt_normal += 1

            deq = deque()
            deq.append([j,i])
            visited[i][j] = True
            while deq:
                cur = deq.popleft()
                for k in range(4):
                    dx = cur[0] + delta_x[k]
                    dy = cur[1] + delta_y[k]
                    if is_valid(dx,dy) and normal_graph[dy][dx] == normal_graph[i][j] and not visited[dy][dx]:
                        deq.append([dx,dy])
                        visited[dy][dx] = True

# 새로 초기화 필수
visited = []
for _ in range(N):
    visited.append([False]*N)

cnt_blind = 0
for i in range(N):
    for j in range(N):
        if not visited[i][j]:
            cnt_blind += 1

            deq = deque()
            deq.append([j,i])
            visited[i][j] = True
            while deq:
                cur = deq.popleft()
                for k in range(4):
                    dx = cur[0] + delta_x[k]
                    dy = cur[1] + delta_y[k]
                    if is_valid(dx,dy) and blind_graph[dy][dx] == blind_graph[i][j] and not visited[dy][dx]:
                        deq.append([dx,dy])
                        visited[dy][dx] = True

print(cnt_normal,cnt_blind)

File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();

    static int n;
    static char[][] graph;
    static boolean[][] visited, visitedD;
    static int[] deltaX = {0, 0, -1, 1};
    static int[] deltaY = {1, -1, 0, 0};

    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(input.readLine());
        graph = new char[n][n];
        for (int i = 0; i < n; i++) {
            String s = input.readLine();
            graph[i] = s.toCharArray();
        }

        visited = new boolean[n][n];
        visitedD = new boolean[n][n];

        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j]) {
                    BFS(i, j, graph[i][j]);
                    cnt++;
                }
            }
        }
        sb.append(cnt).append(" ");

        cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visitedD[i][j]) {
                    BFSD(i, j, graph[i][j]);
                    cnt++;
                }
            }
        }
        sb.append(cnt);
        System.out.println(sb);
    }

    public static void BFS(int y, int x, char c) {
        Deque<int[]> deq = new ArrayDeque<>();
        deq.addLast(new int[]{y, x});
        visited[y][x] = true;
        while (!deq.isEmpty()) {
            int[] cur = deq.pollFirst();
            int curY = cur[0];
            int curX = cur[1];
            for (int i = 0; i < 4; i++) {
                int dx = curX + deltaX[i];
                int dy = curY + deltaY[i];

                if (isValid(dy, dx) && !visited[dy][dx] && graph[dy][dx] == c) {
                    deq.addLast(new int[]{dy, dx});
                    visited[dy][dx] = true;
                }
            }
        }
    }

    public static void BFSD(int y, int x, char c) {
        Deque<int[]> deq = new ArrayDeque<>();
        deq.addLast(new int[]{y, x});
        visitedD[y][x] = true;
        while (!deq.isEmpty()) {
            int[] cur = deq.pollFirst();
            int curY = cur[0];
            int curX = cur[1];
            for (int i = 0; i < 4; i++) {
                int dx = curX + deltaX[i];
                int dy = curY + deltaY[i];

                if (c == 'G' || c == 'R') {
                    if (isValid(dy, dx) && !visitedD[dy][dx] && (graph[dy][dx] == 'G' || graph[dy][dx] == 'R')) {
                        deq.addLast(new int[]{dy, dx});
                        visitedD[dy][dx] = true;
                    }
                } else {
                    if (isValid(dy, dx) && !visitedD[dy][dx] && graph[dy][dx] == c) {
                        deq.addLast(new int[]{dy, dx});
                        visitedD[dy][dx] = true;
                    }
                }
            }
        }
    }

    public static boolean isValid(int x, int y) {
        return x >= 0 && x < n && y >= 0 && y < n;
    }
}